00-Starter-Seed/server.py

```diff
@@ -78,68 +133,62 @@ def requires_scope(required_scope):
+    except jwt.JWTError as ex:
+        raise AuthError({"code": "invalid_token",
+                         "description": "The signature is invalid "}, 401)
+    except Exception:
+        raise AuthError({"code": "invalid_header",
+                         "description":
+                             "Unable to parse authentication"
+                             " token."}, 401)
+    return payload
+
+
 def requires_auth(f):
     """Determines if the access token is valid
     """
     @wraps(f)
     def decorated(*args, **kwargs):
         token = get_token_auth_header()
-        jsonurl = urlopen("https://"+AUTH0_DOMAIN+"/.well-known/jwks.json")
-        jwks = json.loads(jsonurl.read())
-        try:
-            unverified_header = jwt.get_unverified_header(token)
-        except jwt.JWTError:
-            raise AuthError({"code": "invalid_header",
-                            "description":
-                                "Invalid header. "
-                                "Use an RS256 signed JWT Access Token"}, 401)
-        if unverified_header["alg"] == "HS256":
-            raise AuthError({"code": "invalid_header",
-                            "description":
-                                "Invalid header. "
-                                "Use an RS256 signed JWT Access Token"}, 401)
-        rsa_key = {}
-        for key in jwks["keys"]:
-            if key["kid"] == unverified_header["kid"]:
-                rsa_key = {
-                    "kty": key["kty"],
-                    "kid": key["kid"],
-                    "use": key["use"],
-                    "n": key["n"],
-                    "e": key["e"]
-                }
-        if rsa_key:
+
+        rsa_key = cache.get('rsa_key')
+
+        if rsa_key is not None:
             try:
-                payload = jwt.decode(
-                    token,
-                    rsa_key,
-                    algorithms=ALGORITHMS,
-                    audience=AUTH0_AUDIENCE,
-                    issuer="https://"+AUTH0_DOMAIN+"/"
-                )
-            except jwt.ExpiredSignatureError:
-                raise AuthError({"code": "token_expired",
-                                "description": "token is expired"}, 401)
-            except jwt.JWTClaimsError:
-                raise AuthError({"code": "invalid_claims",
-                                "description":
-                                    "incorrect claims,"
-                                    " please check the audience and issuer"}, 401)
-            except Exception:
-                raise AuthError({"code": "invalid_header",
-                                "description":
-                                    "Unable to parse authentication"
-                                    " token."}, 401)
+                jws.verify(token, rsa_key, ALGORITHMS)
+            except jws.JWSError:
+                rsa_key = get_public_key(token)
 
```

Comment:
get_public_key returns it from the cache, right? Can it receive an optional parameter that indicates that we should fetch it anyway?